### PR TITLE
#4 lower shape

### PR DESCRIPTION
**Pull request by chentong319**
Signed-off-by: chentong <chentong@us.ibm.com>

Modified on top of PR#306, which is too old to be merged automatically.

### CODE REVIEW

**Review by AlexandreEichenberger — COMMENTED** [review]
Tong, I would like you to add a test of shape for a memref that has a map that changes the size:

(d0, d1) -> (d0 floordiv 8, d1 floordiv 8, 8, 8)

My understanding is that @doru1004 added the krnl.dim for handling the sizes in presence of such maps. Check your new example, see that the shape remains the logical shape (not the mapped shape). If it does not work, check with lowering to krnl.dim.

**Comment by chentong319** [comment]
@AlexandreEichenberger Should all the lowering code be change to use krnl.dim? If krnl.dim is to ignore the map, it should be test separately, not in lowering of every operations.

File: src/Conversion/ONNXToKrnl/Tensor/Shape.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    // Iterate along the data shape storing dim value to result.
+    for (int i = 0; i < dataRank; i++) {
+      // Create store index value.
+      Value storeVal;
+      Value storeIndex =
+          emitConstantOp(rewriter, loc, rewriter.getIndexType(), i);
+
+      // Checking for dynamic dimensions.
+      if (dataShape[i] == -1) {
+        Value shapeVal = rewriter.create<DimOp>(loc, data, storeIndex);
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToKrnl/Tensor/Shape.cpp:51]: Though we would use the krnl.dim ops

Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/Tensor/Shape.cpp:51]: @chentong319 any action on this one?

Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Shape.cpp:51]: I will push new commit after PR#443 is merged.

Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/Tensor/Shape.cpp:51]: Tong, I though we would use the Krnl.Dim ops instead of the std ones.

Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Shape.cpp]: The new version uses IndexExpr, which, I assume, uses Krnl.Dim.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM

File: src/Conversion/ONNXToKrnl/Tensor/Shape.cpp (hunk 1 above)
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/Tensor/Shape.cpp]: @chentong319  Busted... it is not; will make a patch to correct this.

Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Shape.cpp]: We may need two interfaces: logic dim size or allocated dim size.

**Pull request merged**